lib/fetch.js

```diff
@@ -0,0 +1,88 @@
+import { inflight } from './helpers'
+
+const get = inflight((service, id, params, options) => `${service.path}/${options.queryId}`, getter)
+const find = inflight((service, params, options) => `${service.path}/${options.queryId}`, finder)
+
+export function fetch(feathers, serviceName, method, id, params, { queryId, allPages, parallel }) {
+  const service = feathers.service(serviceName)
+  return method === 'get'
+    ? get(service, id, params, { queryId })
+    : find(service, params, { queryId, allPages, parallel })
+}
+
+function getter(service, id, params) {
+  return service.get(id, params)
+}
+
+function finder(service, params, { queryId, allPages, parallel }) {
+  if (!allPages) {
+    return service.find(params)
+  }
+
+  return new Promise((resolve, reject) => {
+    let skip = 0
+    const result = { data: [], skip: 0 }
+
+    fetchNext()
+
+    function doFind(skip) {
+      return service.find({
+        ...params,
+        query: {
+          ...(params.query || {}),
+          $skip: skip,
+        },
+      })
+    }
+
+    function resolveOrFetchNext(res) {
+      if (res.data.length === 0 || result.data.length >= result.total) {
+        resolve(result)
+      } else {
+        skip = result.data.length
+        fetchNext()
+      }
+    }
+
+    function fetchNextParallel() {
+      const requiredFetches = Math.ceil((result.total - result.data.length) / result.limit)
+
+      if (requiredFetches > 0) {
+        Promise.all(
+          new Array(requiredFetches).fill().map((_, idx) => doFind(skip + idx * result.limit)),
+        )
+          .then(results => {
+            const [lastResult] = results.slice(-1)
+            result.limit = lastResult.limit
+            result.total = lastResult.total
+            result.data = result.data.concat(results.flatMap(r => r.data))
+
+            resolveOrFetchNext(lastResult)
```

Comment:
Normally there would be no more results at this point right? This is if more items were added whilst we were running the query

Reply:
Yes, I think so..